src/rp66.cpp

```diff
@@ -0,0 +1,364 @@
+                const auto msg = "rp66: unexpected EOF when reading header "
+                                 "- got {} bytes";
+                throw protocol_fatal(fmt::format(msg, n));
+            }
+
+
+        default:
+            throw not_implemented(
+                "rp66: unhandled error code in read_header"
+            );
+    }
+
+    // Check the makefile-provided IS_LITTLE_ENDIAN, or the one set by gcc
+    #if (IS_LITTLE_ENDIAN || __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__)
+        std::reverse(b + 0, b + 2);
+    #endif
+
+    header head;
+
+    std::memcpy(&head.length, b + 0, sizeof(head.length));
+    std::memcpy(&head.format, b + 2, sizeof(head.format));
+    std::memcpy(&head.major,  b + 3, sizeof(head.major));
+
+    /*
+     * Flag header as potentially broken if the format version is not 0xFF
+     * 0x01. This is not necessarily a problem, so just flag the header and
+     * continue for now. Should we fail later we have the source in errmsg.
+     *
+     * If the format-version of the current header is correct, it is reasonable
+     * to assume that any issues we flagged on previous headers are non-fatal
+     * and can be ignored.
+     */
+    // TODO: errmsg should be checked elsewhere whenever somethings goes wrong
+    if (head.format != (char)0xFF or head.major != 1) {
+        const auto msg = "rp66: Incorrect format version for header {}";
+        this->fp->errmsg( fmt::format(msg, this->markers.size() + 1) );
+    }
+    else {
+        this->fp->errmsg("");
+    }
+
+    std::int64_t lastbyte = head.length - header::size;
+    if ( this->markers.size() )
+        lastbyte += std::prev(this->markers.end())->lastbyte;
+
+    head.lastbyte = lastbyte;
+
+    this->append(head);
+    this->current = std::prev(this->markers.end());
+    this->current.remaining = head.length - header::size;
+}
+
+
+void rp66::read_header(cursor cur) noexcept (false) {
+    // TODO: Make this a runtime check?
+    assert(this->current.remaining == 0);
+
+    if (std::next(cur) == std::end(this->markers)) {
+        return this->read_header();
+    }
+
+    /*
+     * The record *has* been index'd, so just reposition the underlying stream
+     * and update the internal state
+     */
+    const cursor beg = this->markers.begin();
+    this->current = std::next(cur);
+    const auto rec = std::distance(beg, std::next(this->current));
+
+    const auto tell = (rec * header::size) + cur->lastbyte + this->zero;
+    this->fp->seek(tell);
+    this->current.remaining = this->current->length - header::size;
+}
+
+void rp66::seek_with_index(std::int64_t n) noexcept (false) {
+    std::int64_t records = 1;
+
+    this->current = this->markers.begin();
+    while ( this->current->lastbyte < n ) {
+        records++;
+        this->current++;
+    }
+
+    const std::int64_t real_offset = (records * header::size) + n + this->zero;
+
+    this->fp->seek(real_offset);
+    this->current.remaining = this->current->lastbyte - n;
```

Comment:
Note for future refactoring (not this PR):
- code partly doing the same thing as seek, so could be unified
- for many many VRs we can implement binary search instead of linear.